pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -74,18 +76,103 @@ func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner
 // the pods exceed one node's capacity.
 func (s *Scheduler) schedule(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) []*Node {
 	var nodes []*Node
-	for _, pod := range pods {
-		isScheduled := false
-		for _, node := range nodes {
-			if err := node.Compatible(pod); err == nil {
-				node.Add(pod)
-				isScheduled = true
-				break
+
+	genericInstanceTypes := s.filterGenericInstanceTypes(instanceTypes)
+
+	podSets := s.splitByResourceRequestTypes(pods)
+	for _, ps := range podSets {
+		var newNodes []*Node
+
+		for _, pod := range ps.pods {
+			isScheduled := false
+
+			for _, node := range nodes {
+				if err := node.Compatible(pod); err == nil {
+					node.Add(pod)
+					isScheduled = true
+					break
+				}
+			}
+
+			if !isScheduled {
+				var n *Node
+				if !ps.isExotic {
+					n = NewNode(constraints, genericInstanceTypes, pod)
+				} else {
+					n = NewNode(constraints, instanceTypes, pod)
+				}
+				nodes = append(nodes, n)
+				newNodes = append(newNodes, n)
 			}
 		}
-		if !isScheduled {
-			nodes = append(nodes, NewNode(constraints, instanceTypes, pod))
+
+		// if the pod set contained pods that needed exotic resources, we currently want to prevent any non-exotic
+		// pods from scheduling there as they may increase the instance size unnecessarily.  Once bin-packing and
+		// scheduling are combined, we can just pin the instance type to the lowest one that supports the exotic
+		// workload but still allow scheduling any non-exotic workloads
+		// TODO(todd): when combining bin-packing with scheduling, just pin the instance type instead of preventing scheduling
+		if ps.isExotic {
+			for _, node := range newNodes {
+				node.unschedulable = true
+			}
 		}
 	}
 	return nodes
 }
+
+type podSet struct {
+	pods     []*v1.Pod
+	isExotic bool
```

Comment:
IIUC this abstraction is only necessary to get us over the hump until binpacking and scheduling are combined. FWIW, I think we could probably get away without it, given that the current behavior has this bug. I worry a bit about the complexity or debt that may be left behind from too many code transformations.

Reply:
The issue we seem to keep bumping into with making resource handling more generic is that some resource types  are much more expensive relative to any other types which causes us to a) not want to use instance with those resources if possible, and b) pack pods that do use those resources onto the minimum number of nodes possible.

This code is support "b" where we don't want to create a GPU node for a pod, fill it up with non-GPU workloads, then need to create another GPU node in the same batch.  This problem would also exist when we do bin-packing and scheduling together as far as I can tell.

It's somewhat of a weird thing to worry about though since if the nodes/workloads aren't using tolerations/taints, kube-scheduler will happily schedule non-GPU workloads to the GPU nodes (or similarly to instance types with any other expensive resource in the future).  We're adding complexity to work around a problem that we can't solve outside the single batch of pods we're about to schedule. If workloads are using tolerations/taints to segregate GPU/non-GPU workloads, then part b doesn't exist for us any longer and we can easily handle part a in our AWS cloud provider implementation.

I'm almost convinced we can remove the special resource type registration and just improve the docs to more explicitly state that for expensive resource types you need to use taints/tolerations to prevent co-mingling workloads for both Karpenter and kube-scheduler.  We can then add in a bit of code to the AWS provider to filter out GPU instance types if they aren't required so that capacity optimized selection won't use them, and the problem seems solved.

Reply:
I put a new commit on top that removes all the special registration, leaving special resource handling for taints/tolerations so that kube-scheduler will be aware as well.  This simplifies the code quite a bit and now the AWS provider just has a bit of special code to handle the non-GPU on a GPU instance type issue.